### PR TITLE
Android M support

### PR DESCRIPTION
**Pull request by passsy**
different databases for user specific (with backup) and device specific data

### CODE REVIEW

File: library/src/androidTest/java/net/grandcentrix/tray/storage/TrayStorageTest.java [hunk 1]
```diff
@@ -39,14 +39,15 @@
 
     public void testClear() throws Exception {
         final String MODULE2 = "test2";
-        final TrayStorage storage2 = new TrayStorage(getProviderMockContext(), MODULE2);
+        final TrayStorage storage2 = new TrayStorage(getProviderMockContext(), MODULE2,
+                TrayStorage.Type.USER);
```
Comment (passsy) [inline library/src/androidTest/java/net/grandcentrix/tray/storage/TrayStorageTest.java]: also test this for `Type.DEVICE`

File: library/src/main/java/net/grandcentrix/tray/TrayPreferences.java [hunk 2]
```diff
@@ -39,12 +39,21 @@
...
+    }
+
+    public TrayPreferences(@NonNull final Context context, @NonNull final String module,
             final int version) {
-        super(new TrayStorage(context, module), version);
+        this(context, module, version, TrayStorage.Type.USER);
+    }
+
+    public void annexModule(final String oldStorageName, final TrayStorage.Type type) {
+        super.annex(new TrayStorage(getContext(), oldStorageName, type));
```
Comment (passsy) [inline library/src/main/java/net/grandcentrix/tray/TrayPreferences.java]: add test for annex a module of a different type

File: library/src/main/java/net/grandcentrix/tray/provider/TrayDBHelper.java [hunk 3]
```diff
@@ -34,6 +34,8 @@
 
     public static final String DATABASE_NAME = "tray.db";
 
+    public static final String DATABASE_NAME_NO_BACKUP = "tray_backup_excluded.db";
```
Comment (passsy) [inline library/src/main/java/net/grandcentrix/tray/provider/TrayDBHelper.java]: is `"tray_device_specific"` a better name @jannisveerkamp?

File: library/src/main/java/net/grandcentrix/tray/provider/TrayProvider.java [hunk 4]
```diff
@@ -268,6 +275,17 @@ public int update(final Uri uri, final ContentValues values, final String select
     }
 
     /**
+     * checks the uri for the backup param. default is that
+     *
+     * @param uri contentUri
+     * @return default true or false for {@code /the/uri&backup=false}
+     */
+    private boolean shouldBackup(final Uri uri) {
```
Comment (passsy) [inline library/src/main/java/net/grandcentrix/tray/provider/TrayProvider.java]: I miss a test

File: library/src/main/java/net/grandcentrix/tray/provider/TrayProvider.java [hunk 5]
```diff
@@ -232,7 +239,7 @@ public Cursor query(final Uri uri, final String[] projection, final String selec
 
     @Override
     public void shutdown() {
-        mDbHelper.close();
+        mUserDbHelper.close();
```
Comment (passsy) [inline library/src/main/java/net/grandcentrix/tray/provider/TrayProvider.java]: close `mDeviceDbHelper` too

File: library/src/main/java/net/grandcentrix/tray/provider/TrayUri.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
+package net.grandcentrix.tray.provider;
+
+import net.grandcentrix.tray.storage.TrayStorage.Type;
+
+import android.content.Context;
+import android.net.Uri;
+import android.support.annotation.NonNull;
+
+/**
+ * Created by pascalwelsch on 6/11/15.
```
Comment (passsy) [inline library/src/main/java/net/grandcentrix/tray/provider/TrayUri.java]: javadoc missing

File: library/src/main/java/net/grandcentrix/tray/provider/TrayProvider.java [hunk 7]
```diff
@@ -268,6 +275,17 @@ public int update(final Uri uri, final ContentValues values, final String select
     }
 
     /**
+     * checks the uri for the backup param. default is that
+     *
+     * @param uri contentUri
+     * @return default true or false for {@code /the/uri&backup=false}
```
Comment (passsy) [inline library/src/main/java/net/grandcentrix/tray/provider/TrayProvider.java]: add documentation that `TrayStorage.Type.UNKNOWN` will be backuped

File: library/src/main/java/net/grandcentrix/tray/storage/TrayStorage.java [hunk 8]
```diff
@@ -41,16 +42,29 @@
  */
 public class TrayStorage extends ModularizedStorage<TrayItem> {
 
+    public enum Type {
```
Comment (passsy) [inline library/src/main/java/net/grandcentrix/tray/storage/TrayStorage.java]: javadoc

File: library/src/main/java/net/grandcentrix/tray/storage/TrayStorage.java [hunk 9]
```diff
@@ -72,7 +87,11 @@ public void clear() {
     @Override
     @Nullable
     public TrayItem get(@NonNull final String key) {
-        final Uri uri = mProviderHelper.getUri(getModuleName(), key);
+        final Uri uri = mTrayUri.builder()
+                .setModule(getModuleName())
+                .setType(mType)
+                .setKey(key)
+                .build();
```
Comment (passsy) [inline library/src/main/java/net/grandcentrix/tray/storage/TrayStorage.java]: set type first, than module and key

File: library/src/main/java/net/grandcentrix/tray/storage/TrayStorage.java [hunk 10]
```diff
@@ -151,8 +200,10 @@ public void setVersion(final int version) {
      */
     public void wipe() {
         clear();
-        final Uri uri = mProviderHelper.getInternalUri().buildUpon()
-                .appendPath(getModuleName())
+        final Uri uri = mTrayUri.builder()
+                .setModule(getModuleName())
+                .setType(mType)
+                .setInternal(true)
```
Comment (passsy) [inline library/src/main/java/net/grandcentrix/tray/storage/TrayStorage.java]: internal first

**Comment by passsy** [comment]
A lot more tests are needed. Currently all test use a `User` storage. Using `User` and `Device` storages in parallel is not tested

**Comment by passsy** [comment]
everything working, now I need to focus on coverage and documentation

File: library/src/main/java/net/grandcentrix/tray/provider/TrayDBHelper.java (hunk 3 above)
Comment (jannisveerkamp) [inline library/src/main/java/net/grandcentrix/tray/provider/TrayDBHelper.java]: no - i prefer the way it is :smiley:

**Comment by passsy** [comment]
Here is how I tested the android M auto backup with the sample app:

install sample app and launch it for the first time:


Output of [#testAutoBackup()](https://github.com/grandcentrix/tray/pull/36/files#diff-3a987e1b8f50c7ae6896f15c111d6b80R146)
```
09-14 17:32:55.968    3118-3118/net.grandcentrix.tray.sample V/SampleActivity﹕ deviceId: null
09-14 17:32:55.974    3118-3118/net.grandcentrix.tray.sample V/SampleActivity﹕ no deviceId, created: 894b636d-cf23-4226-9072-f1352a0ad19d
09-14 17:32:55.985    3118-3118/net.grandcentrix.tray.sample V/SampleActivity﹕ userId: null
09-14 17:32:55.996    3118-3118/net.grandcentrix.tray.sample V/SampleActivity﹕ no userId, created: 8b22c88e-fddb-4c6d-8d08-2108b43ea5ea
```


Make backup and restore it. (restore wipes all app data as the first step)

```
Pascals-MBP:~ pascalwelsch$ adb shell
root@generic_x86:/ # bmgr fullbackup net.grandcentrix.tray.sample
Performing full transport backup
root@generic_x86:/ # bmgr restore net.grandcentrix.tray.sample
restoreStarting: 1 packages
onUpdate: 0 = net.grandcentrix.tray.sample
restoreFinished: 0
done
```

Launch sample app

Output of [#testAutoBackup()](https://github.com/grandcentrix/tray/pull/36/files#diff-3a987e1b8f50c7ae6896f15c111d6b80R146)
```
09-14 17:35:09.796    3717-3717/net.grandcentrix.tray.sample V/SampleActivity﹕ deviceId: null
09-14 17:35:09.806    3717-3717/net.grandcentrix.tray.sample V/SampleActivity﹕ no deviceId, created: 253c2d03-75af-4694-92d5-e685d45c591b
09-14 17:35:09.825    3717-3717/net.grandcentrix.tray.sample V/SampleActivity﹕ userId: 8b22c88e-fddb-4c6d-8d08-2108b43ea5ea
```

Worked! The preference with `TrayStorage.Type.USER` was restored, the preference with `TrayStorage.Type.DEVICE` wasn't.


I tested it for both methods: include and exclude (see [code](https://github.com/grandcentrix/tray/blob/feature/android_m_support/sample/src/main/res/xml/mybackupscheme.xml))

**Comment by passsy** [comment]
@jannisveerkamp ready for review! I'll write some documentation

**Comment by passsy** [comment]
I'll add this badge new_badge to the `README.md` :smile:

**Comment by EarlOfEgo** [comment]
:shipit:

**Pull request merged**